File: backend/app/schemas/quote_comparisons.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class ComparisonScenarioInput(BaseModel):
    station_id: uuid.UUID
    product_id: uuid.UUID
    requested_volume_liters: Decimal = Field(..., gt=0)
    comparison_datetime: datetime
    required_delivery_at: datetime | None = None
    ranking_mode: str = "FINANCIAL_EQUIVALENT"
    ranking_scope: str = "BEST_PER_DISTRIBUTOR"


class ReprocessComparisonRequest(BaseModel):
    comparison_datetime: datetime | None = None
    required_delivery_at: datetime | None = None
    requested_volume_liters: Decimal | None = Field(default=None, gt=0)
    ranking_mode: str | None = None
    ranking_scope: str | None = None


class EligibilityReasonResponse(BaseModel):
    code: str
    severity: str
    message: str
    metadata: dict = Field(default_factory=dict)


class ComparisonCostsResponse(BaseModel):
    raw_price_per_liter: Decimal
    discount_per_liter: Decimal
    rebate_per_liter: Decimal
    freight_per_liter: Decimal
    other_cost_per_liter: Decimal
    delivered_cost_per_liter: Decimal
    delivered_total: Decimal
    financial_days: int | None = None
    annual_effective_rate: Decimal | None = None
    daily_rate: Decimal | None = None
    financial_equivalent_cost_per_liter: Decimal | None = None
    financial_equivalent_total: Decimal | None = None


class DistributorBrief(BaseModel):
    id: uuid.UUID
    name: str


class ComparisonResultResponse(BaseModel):
    quote_id: uuid.UUID
    quote_item_id: uuid.UUID
    quote_number: int | None = None
    distributor: DistributorBrief
    eligibility_status: str
    eligibility_reasons: list[EligibilityReasonResponse]
    costs: ComparisonCostsResponse
    rank_position: int | None = None
    difference_per_liter: Decimal | None = None
    difference_total: Decimal | None = None
    is_best_for_distributor: bool = False
    is_best_overall: bool = False
    payment_term_name: str | None = None
    delivery_expected_at: datetime | None = None
    effective_valid_until: datetime | None = None
    calculation_snapshot: dict = Field(default_factory=dict)


class ComparisonSummaryResponse(BaseModel):
    eligible_count: int
    warning_count: int
    ineligible_count: int
    distributor_count: int
    best_cost_per_liter: Decimal | None = None
    highest_cost_per_liter: Decimal | None = None
    average_cost_per_liter: Decimal | None = None
    spread_absolute: Decimal | None = None
    spread_percent: Decimal | None = None


class ComparisonScenarioResponse(BaseModel):
    station_id: uuid.UUID
    product_id: uuid.UUID
    requested_volume_liters: Decimal
    comparison_datetime: datetime
    required_delivery_at: datetime | None
    ranking_mode: str
    ranking_scope: str


class ComparisonRunResponse(BaseModel):
    id: uuid.UUID
    status: str
    methodology_version: str
    scenario: ComparisonScenarioResponse
    summary: ComparisonSummaryResponse
    results: list[ComparisonResultResponse] = Field(default_factory=list)
    calculation_hash: str | None = None
    processing_duration_ms: int | None = None
    reprocessed_from_run_id: uuid.UUID | None = None
    input_snapshot: dict = Field(default_factory=dict)
    created_at: datetime
    created_by: uuid.UUID


class ComparisonRunListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    station_id: uuid.UUID
    product_id: uuid.UUID
    requested_volume_liters: Decimal
    comparison_datetime: datetime
    ranking_mode: str
    ranking_scope: str
    best_cost_per_liter: Decimal | None
    eligible_count: int
    distributor_count: int
    created_at: datetime
    created_by: uuid.UUID


class ComparisonRunListResponse(BaseModel):
    items: list[ComparisonRunListItem]
    total: int
    page: int
    page_size: int

File: frontend/src/api/quote-comparisons.ts
import { apiFetch } from "./client";

export type RankingMode = "RAW" | "DELIVERED" | "FINANCIAL_EQUIVALENT";
export type RankingScope = "BEST_PER_DISTRIBUTOR" | "ALL_OFFERS";

export interface FinancialParameter {
  id: string;
  annual_effective_rate: string;
  day_count_basis: number;
  methodology_version: string;
  valid_from: string;
  valid_until: string | null;
  notes: string | null;
  active: boolean;
}

export interface ComparisonScenarioInput {
  station_id: string;
  product_id: string;
  requested_volume_liters: string;
  comparison_datetime: string;
  required_delivery_at?: string | null;
  ranking_mode?: RankingMode;
  ranking_scope?: RankingScope;
}

export interface EligibilityReason {
  code: string;
  severity: string;
  message: string;
  metadata: Record<string, unknown>;
}

export interface ComparisonCosts {
  raw_price_per_liter: string;
  discount_per_liter: string;
  rebate_per_liter: string;
  freight_per_liter: string;
  other_cost_per_liter: string;
  delivered_cost_per_liter: string;
  delivered_total: string;
  financial_days?: number | null;
  annual_effective_rate?: string | null;
  daily_rate?: string | null;
  financial_equivalent_cost_per_liter?: string | null;
  financial_equivalent_total?: string | null;
}

export interface ComparisonResult {
  quote_id: string;
  quote_item_id: string;
  quote_number?: number | null;
  distributor: { id: string; name: string };
  eligibility_status: string;
  eligibility_reasons: EligibilityReason[];
  costs: ComparisonCosts;
  rank_position?: number | null;
  difference_per_liter?: string | null;
  difference_total?: string | null;
  is_best_for_distributor: boolean;
  is_best_overall: boolean;
  payment_term_name?: string | null;
  delivery_expected_at?: string | null;
  effective_valid_until?: string | null;
  calculation_snapshot: Record<string, unknown>;
}

export interface ComparisonRunListItem {
  id: string;
  station_id: string;
  product_id: string;
  requested_volume_liters: string;
  comparison_datetime: string;
  ranking_mode: RankingMode;
  ranking_scope: RankingScope;
  best_cost_per_liter?: string | null;
  eligible_count: number;
  distributor_count: number;
  created_at: string;
  created_by: string;
}

export interface ComparisonRun {
  id: string;
  status: string;
  methodology_version: string;
  scenario: {
    station_id: string;
    product_id: string;
    requested_volume_liters: string;
    comparison_datetime: string;
    required_delivery_at?: string | null;
    ranking_mode: RankingMode;
    ranking_scope: RankingScope;
  };
  summary: {
    eligible_count: number;
    warning_count: number;
    ineligible_count: number;
    distributor_count: number;
    best_cost_per_liter?: string | null;
    highest_cost_per_liter?: string | null;
    average_cost_per_liter?: string | null;
    spread_absolute?: string | null;
    spread_percent?: string | null;
  };
  results: ComparisonResult[];
  calculation_hash?: string | null;
  created_at: string;
  created_by: string;
}

export async function listFinancialParameters(params?: {
  active?: boolean;
  page?: number;
  page_size?: number;
}) {
  const response = await apiFetch<{ items: FinancialParameter[]; total: number }>(
    `/api/v1/financial-parameters${buildQuery(params)}`,
  );
  return response;
}

function buildQuery(params?: Record<string, string | number | boolean | undefined>) {
  if (!params) return "";
  const query = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== "") query.set(key, String(value));
  });
  const qs = query.toString();
  return qs ? `?${qs}` : "";
}

export async function createFinancialParameter(payload: {
  annual_effective_rate: string;
  day_count_basis?: number;
  valid_from: string;
  valid_until?: string | null;
  notes?: string | null;
}) {
  return apiFetch<FinancialParameter>("/api/v1/financial-parameters", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function runQuoteComparison(payload: ComparisonScenarioInput) {
  return apiFetch<ComparisonRun>("/api/v1/quote-comparisons", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function getQuoteComparison(runId: string) {
  return apiFetch<ComparisonRun>(`/api/v1/quote-comparisons/${runId}`);
}

export async function listQuoteComparisons(params?: Record<string, string | number | undefined>) {
  return apiFetch<{ items: ComparisonRunListItem[]; total: number }>(
    `/api/v1/quote-comparisons${buildQuery(params)}`,
  );
}

export async function getComparisonMethodology() {
  return apiFetch<Record<string, unknown>>("/api/v1/quote-comparisons/methodology");
}

export async function reprocessQuoteComparison(
  runId: string,
  payload: Partial<ComparisonScenarioInput>,
) {
  return apiFetch<ComparisonRun>(`/api/v1/quote-comparisons/${runId}/reprocess`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function comparisonExportUrl(runId: string, format: "pdf" | "csv") {
  return `/api/v1/quote-comparisons/${runId}/export/${format}`;
}
